devel/api.md

```diff
@@ -0,0 +1,990 @@
+    * [`Route.DESTINATION`](#routedestination)
+    * [`Route.NEXT_HOP_INTERFACE`](#routenext_hop_interface)
+    * [`Route.NEXT_HOP_ADDRESS`](#routenext_hop_address)
+    * [`Route.METRIC`](#routemetric)
+* [DNS client configuration](#dns-client-configuration)
+    * [`DNS.SEARCH`](#dnssearch)
+    * [`DNS.SERVER`](#dnsserver)
+
+<!-- vim-markdown-toc -->
+
+# Introduction
+
+This document holds all the API detail of the `libnmstate` Python module.
+For quick start, you may refer to [Python code examples][py_example]
+
+The `libnmstate` contains:
+ * Functions:
+    * `show()`  -- [Query network state][show]
+    * `apply()` -- [Apply network state][apply]
+    * `commit()` -- [Commit a transaction][manual]
+    * `rollback()` -- [Rollback a transaction][manual]
+ * [All exception classes][error] are stored in `libnmstate.error` and
+   are diverted from `libnmstate.error.NmstateError`.
+ * All other classes holding constants are stored in `libnmstate.schema`
+   module.
+
+## Query network state
+
+Example of querying network state:
+
+```python
+import json
+import libnmstate
+
+net_state = libnmstate.show()
+print(json.dumps(net_state, indent=4))
+```
+
+## Apply network state
+
+Nmstate supports partial editing, only the specified data is changed, the
+unspecified data is preserved/untouched.
+
+Example of changing eth1 MTU to 1460:
+
+```python
+import libnmstate
+from libnmstate.schema import Interface
+
+libnmstate.apply({
+    Interface.KEY: [
+        {
+            Interface.NAME: 'eth1',
+            Interface.MTU: 1460
+        }
+    ]
+})
+```
+
+## Manual Transaction Control
+
+## Error handling
+
+All the exceptions are stored in `libnmstate.error` and are all
+is inherited from `NmstateError`.
+
+### NmstateError
+
+The base exception of libnmstate.
+
+### NmstateDependencyError
+
+Nmstate requires external tools installed and/or started for desired state.
+
+### NmstateValueError
+
+Exception happens at pre-apply check, user should resubmit the amended
+desired state. Example:
+    * JSON/YAML syntax issue.
+    * Nmstate schema issue.
+    * Invalid value of desired property, like bond missing slave.
+
+### NmstatePermissionError
+
+Permission deny when applying the desired state.
+
+
+### NmstateConflictError
+
+Something else is already editing the network state via Nmstate.
+
+
+### NmstateLibnmError
+
+Exception for unexpected NetworkManager libnm failure.
+
+
+### NmstateVerificationError
+
+After applied desired state, current state does not match desired state for
+unknown reason.
+
+
+### NmstateNotImplementedError
+
+Desired feature is not supported by Nmstate yet.
+
+### NmstateInternalError
+
+Unexpected behaviour happened. It is a bug of libnmstate which should be fixed.
+
+## Schema
+
+The YAML schema file is placed at `libnmstate/schemas/operational-state.yaml`.
+
+In stead of using hard coded string, please use classes and constants in
+`libnmstate.schmea`.
+
+# Limitations
```

Comment:
I think this belongs to a separate file in the source repo. Changing the user guide when a bug is fixed is not that common.
But we could place a link here to the limitation text.

Reply:
https://github.com/nmstate/nmstate/pull/500